### PR TITLE
Fix `torchmetrics` compatibility

### PR DESCRIPTION
**Pull request by Borda**
## What does this PR do?

fix some compatibility issues

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -8,5 +8,5 @@ PyYAML>=5.1, !=5.4.*  # OmegaConf requirement >=5.1
 tqdm>=4.41.0
 fsspec[http]>=0.8.1
 tensorboard>=2.2.0
-torchmetrics>=0.2.0
+torchmetrics>=0.3.0
```
Comment (awaelchli) [inline requirements.txt:11]: we could just pin == to avoid it in the future

**Review by awaelchli — APPROVED** [review]

File: requirements.txt (hunk 1 above)
Comment (Borda) [reply requirements.txt:11]: that was my first thought too but that would block any user using newer TM

Comment (awaelchli) [reply requirements.txt:11]: don't we want users to import torchmetrics directly?

Comment (Borda) [reply requirements.txt:11]: yes but still we need to handle compatibility

File: pytorch_lightning/metrics/functional/iou.py [hunk 2]
```diff
@@ -21,7 +21,7 @@
 
 @deprecated_metrics(target=_iou)
 def iou(
-    pred: torch.Tensor,
+    preds: torch.Tensor,
```
Comment (Borda) [inline pytorch_lightning/metrics/functional/iou.py:24]: this will be handled with conditional wrapper https://github.com/Borda/pyDeprecate/pull/4

**Review by Borda — COMMENTED** [review]
releasing updated wrapper which shall handle this dis mapping https://github.com/Borda/pyDeprecate/pull/4

File: pytorch_lightning/metrics/classification/f_beta.py [hunk 3]
```diff
@@ -17,11 +17,13 @@
 from torchmetrics import FBeta as _FBeta
 
 from pytorch_lightning.metrics.utils import deprecated_metrics
+from pytorch_lightning.utilities.imports import _TORCHMETRICS_GREATER_EQUAL_0_3, _TORCHMETRICS_LOWER_THAN_0_3
 
 
 class FBeta(_FBeta):
 
-    @deprecated_metrics(target=_FBeta)
+    @deprecated_metrics(target=_FBeta, skip_if=_TORCHMETRICS_GREATER_EQUAL_0_3)
```
Comment (Borda) [inline pytorch_lightning/metrics/classification/f_beta.py:25]: @justusschock

File: pytorch_lightning/metrics/classification/f_beta.py [hunk 4]
```diff
@@ -17,11 +17,13 @@
...
+    @deprecated_metrics(target=_FBeta, args_mapping=dict(multilabel="multiclass"), skip_if=_TORCHMETRICS_LOWER_THAN_0_3)
```
Comment (Borda) [inline pytorch_lightning/metrics/classification/f_beta.py:26]: @justusschock

File: .github/workflows/ci_test-conda.yml [hunk 5]
```diff
@@ -31,6 +31,8 @@ jobs:
         python ./requirements/adjust_versions.py requirements/extra.txt
         python ./requirements/adjust_versions.py requirements/examples.txt
         pip install --requirement requirements/devel.txt --upgrade-strategy only-if-needed
+        # TODO: remove later this is just to test to TM fix
+        pip install https://github.com/PyTorchLightning/metrics/archive/refs/heads/fix/PL-compatible.zip
```
Comment (Borda) [inline .github/workflows/ci_test-conda.yml:35]:
```suggestion
        pip install https://github.com/PyTorchLightning/metrics/archive/refs/heads/master.zip
```

File: .github/workflows/ci_test-full.yml [hunk 6]
```diff
@@ -135,6 +135,8 @@ jobs:
         python ./requirements/adjust_versions.py requirements/extra.txt
         python ./requirements/adjust_versions.py requirements/examples.txt
         pip install --requirement ./requirements/devel.txt --find-links https://download.pytorch.org/whl/cpu/torch_stable.html --upgrade
+        # TODO: remove later this is just to test to TM fix
+        pip install https://github.com/PyTorchLightning/metrics/archive/refs/heads/fix/PL-compatible.zip
```
Comment (Borda) [inline .github/workflows/ci_test-full.yml:139]:
```suggestion
        pip install https://github.com/PyTorchLightning/metrics/archive/refs/heads/master.zip
```

File: pytorch_lightning/metrics/classification/precision_recall.py [hunk 7]
```diff
@@ -71,3 +71,5 @@ def __init__(
         .. deprecated::
             Use :class:`~torchmetrics.Recall`. Will be removed in v1.5.0.
         """
+        _ = num_classes, threshold, average, multilabel, mdmc_average, ignore_index, top_k, is_multiclass, \
+            compute_on_step, dist_sync_on_step, process_group, dist_sync_fn
```
Comment (Borda) [inline pytorch_lightning/metrics/classification/precision_recall.py:75]:
```suggestion
```

**Review by ananthsub — APPROVED** [review]

File: .github/workflows/ci_test-full.yml [hunk 8]
```diff
@@ -135,6 +135,8 @@ jobs:
         python ./requirements/adjust_versions.py requirements/extra.txt
         python ./requirements/adjust_versions.py requirements/examples.txt
         pip install --requirement ./requirements/devel.txt --find-links https://download.pytorch.org/whl/cpu/torch_stable.html --upgrade
+        # TODO: remove later this is just to test to TM fix
+        pip install https://github.com/PyTorchLightning/metrics/archive/refs/heads/master.zip
```
Comment (Borda) [inline .github/workflows/ci_test-full.yml:139]:
```suggestion
```

**Comment by carmocca** [comment]
@Borda Docker failures are real

**Review by awaelchli — APPROVED** [review]

**Comment by Borda** [comment]
> @Borda Docker failures are real

that is some strange cache because it complains about
```
TypeError: deprecated() got an unexpected keyword argument 'skip_if'
```
but requirement set `pyDeprecate==0.3.0` and it is on pypi https://pypi.org/project/pyDeprecate/0.3.0

**Review by justusschock — APPROVED** [review]

**Comment by carmocca** [comment]
> TypeError: deprecated() got an unexpected keyword argument 'skip_if'
> ```
> 
> but requirement set `pyDeprecate==0.3.0` and it is on pypi https://pypi.org/project/pyDeprecate/0.3.0

Still broken :/
https://dev.azure.com/PytorchLightning/pytorch-lightning/_build/results?buildId=6322&view=logs&j=3afc50db-e620-5b81-6016-870a6976ad29&t=25e9fc0a-700e-5efd-103d-dbf01d2cf0ab&l=206

**Comment by Borda** [comment]
> Still broken :/

that is fishy as the very same cal pass in standard tests in the very same workflow...

**Review by tchaton — APPROVED** [review]
LGTM ! Thanks !

**Pull request merged**